src/inspector/worker_agent.cc

```diff
@@ -0,0 +1,155 @@
+#include "worker_agent.h"
+
+#include "main_thread_interface.h"
+#include "worker_inspector.h"
+
+namespace node {
+namespace inspector {
+namespace protocol {
+
+class NodeWorkers
+    : public std::enable_shared_from_this<NodeWorkers> {
+ public:
+  explicit NodeWorkers(std::weak_ptr<NodeWorker::Frontend> frontend,
+                      std::shared_ptr<MainThreadHandle> thread)
+                      : frontend_(frontend), thread_(thread) {}
+  void WorkerCreated(const std::string& title,
+                 const std::string& url,
+                 bool waiting,
+                 std::shared_ptr<MainThreadHandle> target);
+  void Receive(const std::string& id, const std::string& message);
+  void Send(const std::string& id, const std::string& message);
+  void Detached(const std::string& id);
+
+ private:
+  std::weak_ptr<NodeWorker::Frontend> frontend_;
+  std::shared_ptr<MainThreadHandle> thread_;
+  std::unordered_map<std::string, std::unique_ptr<InspectorSession>> sessions_;
+  int next_target_id_ = 0;
+};
+
+namespace {
+class AgentWorkerInspectorDelegate : public WorkerDelegate {
+ public:
+  explicit AgentWorkerInspectorDelegate(std::shared_ptr<NodeWorkers> workers)
+                                        : workers_(workers) {}
+
+  void WorkerCreated(const std::string& title,
+                     const std::string& url,
+                     bool waiting,
+                     std::shared_ptr<MainThreadHandle> target) override {
+    workers_->WorkerCreated(title, url, waiting, target);
+  }
+
+ private:
+  std::shared_ptr<NodeWorkers> workers_;
+};
+
+class ParentInspectorSessionDelegate : public InspectorSessionDelegate {
+ public:
+  ParentInspectorSessionDelegate(const std::string& id,
+                                 std::shared_ptr<NodeWorkers> workers)
+                                 : id_(id), workers_(workers) {}
+
+  ~ParentInspectorSessionDelegate() override {
+    workers_->Detached(id_);
+  }
+
+  void SendMessageToFrontend(const v8_inspector::StringView& msg) override {
+    std::string message = protocol::StringUtil::StringViewToUtf8(msg);
+    workers_->Send(id_, message);
+  }
+
+ private:
+  std::string id_;
+  std::shared_ptr<NodeWorkers> workers_;
+};
+
+std::unique_ptr<NodeWorker::WorkerInfo> WorkerInfo(const std::string& id,
+                                                   const std::string& title,
+                                                   const std::string& url) {
+  return NodeWorker::WorkerInfo::create()
+      .setWorkerId(id)
+      .setTitle(title)
+      // Placeholder URL still needed for some frontends to resolve paths
+      .setUrl(url.empty() ? "file://eval" : url)
```

Comment:
Is it intentional that this has 2 slashes, i.e. the *host* is being set to `eval`?

Reply:
That's what Chrome DevTools team suggested, but they do not have equivalent feature in Chrome. It might be ok to set it to another value.

Reply:
> That's what Chrome DevTools team suggested

Hm. I would simply pass url as is. Empty is empty, that should be fine.

Reply:
Empty URL it is.